src/main/java/org/stratumproject/fabric/tna/behaviour/upf/FabricUpfProgrammable.java

```diff
@@ -547,6 +550,26 @@ private Collection<UpfEntity> getUpfTerminationsDownlink() throws UpfProgrammabl
         return upfTerminations;
     }
 
+    private void applyUpfCounter(UpfCounter upfCounter) throws UpfProgrammableException {
+        PiEntity piIngressCounter = new PiCounterCell(
+                PiCounterCellId.ofIndirect(FABRIC_INGRESS_UPF_TERMINATIONS_COUNTER, upfCounter.getCellId()),
+                new PiCounterCellData(upfCounter.getIngressPkts(), upfCounter.getIngressBytes()));
+        PiEntity piEgressCounter = new PiCounterCell(
+                PiCounterCellId.ofIndirect(FABRIC_EGRESS_UPF_TERMINATIONS_COUNTER, upfCounter.getCellId()),
+                new PiCounterCellData(upfCounter.getEgressPkts(), upfCounter.getEgressBytes()));
+        // Write on the device in sync mode.
+        P4RuntimeWriteClient.WriteResponse writeResponse = client.write(
+                        DEFAULT_P4_DEVICE_ID, pipeconf)
+                .modify(Lists.newArrayList(piIngressCounter, piEgressCounter))
+                .submitSync();
+        if (!writeResponse.failed().isEmpty()) {
+            writeResponse.failed().stream().filter(counterEntryResp -> !counterEntryResp.isSuccess())
+                    .forEach(counterEntryResp -> log.error("A counter was not modified correctly: {}",
+                                                          counterEntryResp.explanation()));
+            throw new UpfProgrammableException("Failed to modify counters");
```

Comment:
Is this a severe failure ? Should we drop the exception ?

Reply:
Generating the exception is the only way to convey a failure to the caller. If the counter wasn't modified I would prefer to know it in the caller and eventually propagating it to pfcp-agent, otherwise we lose this failure.
In general, anyway, I think we should remove the generic UpfCounter and have separate ingress and egress counters.

Reply:
This might be out of scope - I was curious about the behavior of other `apply` type and realize most of them doesn't actually throw exceptions when flow rule apply fails. Is there a reason that we don't utilize exception as a signal of failure in those cases?

Reply:
The main difference is that all other applies interact with ONOS subsystems (i.e., MeterService or FlowRuleService), meaning that for UP4 perspective, as long as the entities has been correctly translated and inserted in we are ok. In those cases, we propagate mainly translation errors. For counters, it's a little bit different, since we don't have an actual ONOS subsystem, we have to directly use the p4runtime client to interact with the device, that's the reason why we propagate the exception.

Reply:
nvm - now it is clear the scenario